edb/server/protocol/auth_ext/base.py

```diff
@@ -16,18 +16,129 @@
 # limitations under the License.
 #
 
+import immutables
+import base64
+import datetime
+import http
 
-class BaseProvider:
-    def __init__(self, name: str, client_id: str, client_secret: str):
+from typing import Type, TypeVar, overload
+from jwcrypto import jwk, jwt
+
+from edb import errors
+from edb.server.config import ops
+
+SettingsMap = immutables.Map[str, ops.SettingValue]
+
+
+class BaseOAuthProvider:
+    def __init__(
+        self,
+        name: str,
+        db_config: SettingsMap,
+    ):
+        client_id, client_secret = _get_client_credentials(name, db_config)
+        signing_key = _get_auth_signing_key(db_config)
         self.name = name
         self.client_id = client_id
         self.client_secret = client_secret
+        self.signing_key = signing_key
+
+    async def authorize(
+        self, response, redirect_uri: str, issuer: str
+    ):
+        state_token = self._make_signed_token(iss=issuer, provider=self.name)
+
+        auth_url = self._get_code_url(
+            state=state_token.serialize(), redirect_uri=redirect_uri
+        )
+
+        response.status = http.HTTPStatus.FOUND
+        response.custom_headers["Location"] = auth_url
+        response.close_connection = True
```

Comment:
This _seems_ generic enough that I moved it into the base class, but I'd be curious to know if it doesn't seem right. I think this roughly matches the separation we had with the separate OAuth module where the `Provider` knows how to build the auth-provider-specific redirect URL, but the OAuth logic is generalized.